samples/gibbs_ensemble/gibbs.py

```diff
@@ -0,0 +1,139 @@
+import socket
+import struct
+import pickle
+import numpy as np
+
+
+class MessageId:
+    START = 0
+    END = 1
+    MOVE_PART = 2
+    MOVE_PART_REVERT = 21
+    CHANGE_VOLUME = 3
+    CHANGE_VOLUME_REVERT = 31
+    EXCHANGE_PART_ADD = 4
+    EXCHANGE_PART_ADD_REVERT = 41
+    EXCHANGE_PART_REMOVE = 5
+    EXCHANGE_PART_REMOVE_REVERT = 51
+    ENERGY = 6
+
+
+class Client():
+    '''
+    Base class for a Gibbs ensemble client. Each client runs a single
+    system, executes requested changes and returns energies.
+    '''
+    _system = None
+
+    def __init__(self, system):
+        self._system = system
+
+    def random_particle(self):
+        id = np.random.choice(self._system.part[:].id)
+        return self._system.part[id]
+
+    def move_particle(self):
+        '''Moves a particle to a new random position. The old position is saved in _old_part_position.'''
+        p = self.random_particle()
+        self._last_modified_pid = p.id
+        self._saved_pos = p.pos
+        p.pos = np.random.random(3) * self._system.box_l
+
+    def move_particle_revert(self):
+        '''Revert the last movement step.'''
+        self._system.part[self._last_modified_pid].pos = self._saved_pos
+
+    def backup_particle(self, p):
+        """This implementation only saves position"""
+        self._saved_particle_state = {'id': p.id, 'pos': p.pos}
+
+    def remove_particle(self):
+        '''Removes a random particle. The old position and index are saved'''
+        p = self.random_particle()
+        self.backup_particle(p)
+        p.remove()
```

Comment:
if we wanted to show best practices here, we should probably not use a lot of function with side-effects and consider changing it to a more functional style approach with free functions that return stuff (and can be tested)